cmd/main_test.go

```diff
@@ -0,0 +1,220 @@
+	recorder := httptest.NewRecorder()
+	probeHandler(recorder, req, log.NewNopLogger(), config.Config{})
+
+	resp := recorder.Result()
+	body, _ := ioutil.ReadAll(resp.Body)
+
+	if resp.StatusCode != http.StatusOK {
+		t.Fatalf("Default 'Accept: application/json' header test fails unexpectedly, got %s", body)
+	}
+}
+
+func TestCorrectResponse(t *testing.T) {
+	tests := []struct {
+		ConfigFile    string
+		ServeFile     string
+		ResponseFile  string
+		ShouldSucceed bool
+	}{
+		{"../test/config/good.yml", "/serve/good.json", "../test/response/good.txt", true},
+		{"../test/config/good.yml", "/serve/repeat-metric.json", "../test/response/good.txt", false},
+	}
+
+	target := httptest.NewServer(http.FileServer(http.Dir("../test")))
+	defer target.Close()
+
+	for i, test := range tests {
+		c, err := config.LoadConfig(test.ConfigFile)
+		if err != nil {
+			t.Fatalf("Failed to load config file %s", test.ConfigFile)
+		}
+
+		req := httptest.NewRequest("GET", "http://example.com/foo"+"?target="+target.URL+test.ServeFile, nil)
+		recorder := httptest.NewRecorder()
+		probeHandler(recorder, req, log.NewNopLogger(), c)
+
+		resp := recorder.Result()
+		body, _ := ioutil.ReadAll(resp.Body)
+
+		expected, _ := ioutil.ReadFile(test.ResponseFile)
+
+		if test.ShouldSucceed && string(body) != string(expected) {
+			t.Fatalf("Correct response validation test %d fails unexpectedly.\nGOT:\n%s\nEXPECTED:\n%s", i, body, expected)
+		}
+	}
+}
+
+func TestBasicAuth(t *testing.T) {
+	username := "myUser"
+	password := "mySecretPassword"
+	expected := "Basic " + base64.StdEncoding.EncodeToString([]byte(username+":"+password))
+	target := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if got := r.Header.Get("Authorization"); got != expected {
+			t.Errorf("BasicAuth mismatch, got: %s, expected: %s", got, expected)
+			w.WriteHeader(http.StatusUnauthorized)
+		}
+	}))
+	defer target.Close()
+
+	req := httptest.NewRequest("GET", "http://example.com/foo"+"?target="+target.URL, nil)
+	recorder := httptest.NewRecorder()
+	c := config.Config{}
+	auth := &pconfig.BasicAuth{
+		Username: username,
+		Password: pconfig.Secret(password),
+	}
+
+	c.HTTPClientConfig.BasicAuth = auth
+	probeHandler(recorder, req, log.NewNopLogger(), c)
+
+	resp := recorder.Result()
+	body, _ := ioutil.ReadAll(resp.Body)
+
+	if resp.StatusCode != http.StatusOK {
+		t.Fatalf("BasicAuth test fails unexpectedly. Got: %s", body)
+	}
+}
+
+func TestBearerToken(t *testing.T) {
+	token := "mySecretToken"
+	target := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		expected := "Bearer " + token
+		if got := r.Header.Get("Authorization"); got != expected {
+			t.Errorf("BearerToken mismatch, got: %s, expected: %s", got, expected)
+			w.WriteHeader(http.StatusUnauthorized)
+		}
+	}))
+	defer target.Close()
+
+	req := httptest.NewRequest("GET", "http://example.com/foo"+"?target="+target.URL, nil)
```

Comment:
That would not protect the URL correctly if it has & and ?

Maybe url.QueryEscape(target.URL)?

Reply:
Since this is only in the unit test it will always be `http://127.0.0.1:<random_port>` or `http://[::1]:<random_port>`. If any complex unit tests are added in future which make use of the `&` or `?` characters in the destination url for fetching json data, then escaping the destination only in that unit test would make more sense I think.